docs/rl-algorithms/ddpg.md

```diff
@@ -0,0 +1,195 @@
+# Deep Deterministic Policy Gradient (DDPG)
+
+
+## Overview
+
+DDPG is a popular DRL algorithm for continuous control. It runs reasonably fast by leveraging vector (parallel) environments and naturally works well with different action spaces, therefore supporting a variety of games. It also has good sample efficiency compared to algorithms such as DQN.
+
+
+Original paper: 
+
+* [Continuous control with deep reinforcement learning](https://arxiv.org/abs/1509.02971)
+
+Reference resources:
+
+* :material-github: [sfujim/TD3](https://github.com/sfujim/TD3)
+
+## Implemented Variants
+
+
+| Variants Implemented      | Description |
+| ----------- | ----------- |
+| :material-github: [`ddpg_continuous_action.py`](https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ddpg_continuous_action.py), :material-file-document: [docs](/rl-algorithms/ddpg/#ddpg_continuous_actionpy) | For continuous action space. Also implemented Mujoco-specific code-level optimizations |
+
+
+Below are our single-file implementations of PPO:
+
+## `ddpg_continuous_action.py`
+
+The [ddpg.py](https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ddpg.py) has the following features:
+
+* For continuous action space. Also implemented Mujoco-specific code-level optimizations
+* Works with the `Box` observation space of low-level features
+* Works with the `Box` (continuous) action space
+
+### Usage
+
+```bash
+poetry install
+poetry install -E pybullet
+python cleanrl/ddpg_continuous_action.py --help
+python cleanrl/ddpg_continuous_action.py --env-id HopperBulletEnv-v0
+poetry install -E mujoco # only works in Linux
+python cleanrl/ddpg_continuous_action.py --env-id Hopper-v3
+```
+
+### Explanation of the logged metrics
+
+Running `python cleanrl/ddpg_continuous_action.py` will automatically record various metrics such as various losses in Tensorboard. Below are the documentation for these metrics:
+
+* `charts/episodic_return`: episodic return of the game
+* `charts/SPS`: number of steps per second
+* `losses/qf1_loss`: the MSE between the Q values at timestep $t$ and the target Q values at timestep $t+1$, which minimizes temporal difference. 
+* `losses/actor_loss`: implemented as `-qf1(data.observations, actor(data.observations)).mean()`; it is the *negative* average Q values calculated based on the 1) observations and the 2) actions computed by the actor based on these observations. By minimizing `actor_loss`, the optimizer updates the actors parameter using the following gradient (Lillicrap et al., 2016, Equation 6)[^1]:
+
+$$ \mathbb{E}_{s_{t} \sim \rho^{\beta}}\left[\left.\left.\nabla_{a} Q\left(s, a \mid \theta^{Q}\right)\right|_{s=s_{t}, a=\mu\left(s_{t}\right)} \nabla_{\theta_{\mu}} \mu\left(s \mid \theta^{\mu}\right)\right|_{s=s_{t}}\right]$$
```

Comment:
What's the $$\rho^{\beta}$$? Are u referring the buffer $$ R $$?

Reply:
$$\rho^{\beta}$$ has a definition as follows

<img width="534" alt="image" src="https://user-images.githubusercontent.com/5555347/159189804-e23d87d5-011c-4b79-b969-17343be4e00a.png">

I didn't add too much explanation for this because I referenced equation 6 from the original paper. If you think this could confuse the readers, I can remove it and just say equation 6 without copying it in our docs.

Reply:
But the equation 6 is DPG's update method.
I think DDPG uses $$ \frac{1}{N}\sum_i\left.\left.\nabla_{a} Q\left(s, a \mid \theta^{Q}\right)\right|{s=s_{i}, a=\mu\left(s_{i}\right)} \nabla_{\theta_{\mu}} \mu\left(s \mid \theta^{\mu}\right)\right|{s_{i}} $$
as shown in
![image](https://user-images.githubusercontent.com/19211761/159190601-34a5ba50-948c-42c1-8eda-ed02f91497cd.png)

Reply:
You're absolutely right. Thanks for the couch and I have just fixed it.